### PR TITLE
Core: Fix equality in StructLikeMap

### PR DESCRIPTION
**Pull request by aokolnychyi**
This PR fixes equality in `StructLikeMap`. See tests for use cases that previously failed.

### CODE REVIEW

File: core/src/main/java/org/apache/iceberg/util/StructLikeMap.java [hunk 1]
```diff
@@ -146,25 +145,19 @@ public R getValue() {
 
     @Override
     public int hashCode() {
-      int hashCode = getKey().hashCode();
```
Comment (aokolnychyi) [inline core/src/main/java/org/apache/iceberg/util/StructLikeMap.java:149]: This was also throwing NPE cause key could be null. Not anymore.

File: core/src/main/java/org/apache/iceberg/util/StructLikeMap.java [hunk 2]
```diff
@@ -146,25 +145,19 @@ public R getValue() {
...
-    @SuppressWarnings("unchecked")
     public boolean equals(Object o) {
       if (this == o) {
         return true;
-      } else if (!(o instanceof StructLikeEntry)) {
+      } else if (o == null || getClass() != o.getClass()) {
         return false;
-      } else {
-        StructLikeEntry that = (StructLikeEntry<R>) o;
-        return Objects.equals(getKey(), that.getKey())
```
Comment (aokolnychyi) [inline core/src/main/java/org/apache/iceberg/util/StructLikeMap.java:165]: We used `getKey()` which unwrapped the object and we were loosing proper equality.

File: core/src/test/java/org/apache/iceberg/util/TestStructLikeMap.java [hunk 3]
```diff
@@ -147,4 +149,32 @@ public void testKeysWithNulls() {
 
     assertThat(map.remove(record3)).isEqualTo("aaa");
   }
+
+  @Test
+  public void testEqualsAndHashCode() {
```
Comment (RussellSpitzer) [inline core/src/test/java/org/apache/iceberg/util/TestStructLikeMap.java:154]: Can we do a quick addition of a test for empty map equality? I am pretty sure that is fine but just to be complete

Comment (aokolnychyi) [reply core/src/test/java/org/apache/iceberg/util/TestStructLikeMap.java:154]: Sure, I'll add that.

**Comment by aokolnychyi** [comment]
Thanks, @RussellSpitzer!

**Pull request merged**